core/src/main/java/org/hisp/dhis/android/core/trackedentity/TrackedEntityAttributeReservedValueManager.kt

```diff
@@ -0,0 +1,407 @@
+    }
+
+    /**
+     * Get a reserved value and remove it from database. If the number of available values is below a threshold
+     * (default [.FILL_UP_TO] * [.FACTOR_TO_REFILL]) it tries to download before returning a value.
+     *
+     * @param attributeUid        Attribute uid
+     * @param organisationUnitUid Organisation unit uid
+     * @return Single with value of tracked entity attribute
+     */
+
+    fun getValue(attributeUid: String, organisationUnitUid: String): Single<String> {
+        return rxSingle { getValueCoroutines(attributeUid, organisationUnitUid) ?: "" }
+    }
+
+    private suspend fun getValueCoroutines(attributeUid: String, organisationUnitUid: String): String {
+
+        downloadValuesIfBelowThreshold(attributeUid, getOrganisationUnit(organisationUnitUid), null, false)
+
+        val pattern = trackedEntityAttributeStore.selectByUid(attributeUid)!!.pattern()
+        val attributeOrgunit = if (isOrgunitDependent(pattern)) organisationUnitUid else null
+
+        return store.popOne(attributeUid, attributeOrgunit)?.value() ?: throw D2Error.builder()
+            .errorCode(D2ErrorCode.NO_RESERVED_VALUES).errorDescription("There are no reserved values")
+            .errorComponent(D2ErrorComponent.Database).build()
+    }
+
+    /**
+     * @param attributeUid           Attribute uid
+     * @param numberOfValuesToFillUp An optional maximum number of values to reserve
+     * @see .downloadReservedValues
+     */
+    fun blockingDownloadReservedValues(
+        attributeUid: String,
+        numberOfValuesToFillUp: Int?
+    ) {
+        runBlocking { downloadReservedValues(attributeUid, numberOfValuesToFillUp) }
+    }
+
+    /**
+     * Download of TrackedEntityInstance reserved values. The number of reserved values is filled up to the
+     * numberOfValuesToFillUp. If not defined, it defaults to [.FILL_UP_TO].
+     * <br></br><br></br>
+     * The download is only triggered for this attribute passed as parameter.
+     * <br></br><br></br>
+     * If the attribute pattern is dependent on OrganisationUnit code (that is, it contains "ORG_UNIT_CODE"), it
+     * reserves values for each orgunit assigned to the programs with this attribute. It applies the limit
+     * per orgunit. Otherwise the limit is applied per attribute.
+     *
+     * @param attributeUid           Attribute uid
+     * @param numberOfValuesToFillUp An optional maximum number of values to reserve
+     * @return An Observable that notifies about the progress.
+     */
+
+    fun downloadReservedValues(
+        attributeUid: String,
+        numberOfValuesToFillUp: Int?
+    ): Observable<D2Progress> {
+        return downloadReservedValuesFlow(attributeUid, numberOfValuesToFillUp).asObservable()
+    }
+
+    private fun downloadReservedValuesFlow(
+        attributeUid: String,
+        numberOfValuesToFillUp: Int?
+    ) = flow<D2Progress> {
+        downloadValuesForOrgUnits(attributeUid, numberOfValuesToFillUp).collect()
```

Comment:
I think this flow does not emit anything. It should call `emit` or `emitAll`, something like 
`emitAll(downloadValuesForOrgUnits(attributeUid, numberOfValuesToFillUp))`

Reply:
I think these are just two ways of starting the flow, maybe `emitAll` looks more clean, but they should do the same